### PR TITLE
My Solution

### PR DESCRIPTION
**Pull request by agus24**
- create `sendJsonResponse` and `sendErrorResponse` inside base controller so it can be easily use
- Using gate / policies
- use relation to check if user is voted (so i dont need to check `question_id` anymore)
- change relation name from `voice` to `voices` (i think it should be voices because it's one to many relationship)
- create `VoiceRequest` to remove `$request->validate(...)` from controller
- change some error message to make it easier for user to understand

### CODE REVIEW

File: Controller.php [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    }
+
+    /**
+     * Send JSON Error Response
+     *
+     * @param string $message
+     * @param integer $code
+     * @return Response
+     */
+    protected function sendErrorResponse(string $message, int $code): Response
```
Comment (efalsett) [inline Controller.php:31]: Some could say `$code` could have a default value with 200 for a normal response and some 400ish for error response.

Comment (agus24) [reply Controller.php:31]: should i make just 1 type of response method then check if `code < 400` then send normal response else send error response?

File: VoiceController.php [hunk 2]
```diff
@@ -1,48 +1,38 @@
...
+            return $this->sendErrorResponse("You cant vote to your own question", 403);
+        }
 
-    return response()->json([
-        'status'=>200,
-        'message'=>'Voting completed successfully'
-    ]);
-}
\ No newline at end of file
+        $voice = $question->voices()->where('user_id', auth()->id())->first();
```
Comment (xmokko) [inline VoiceController.php:24]: $question->voices() - generate another SQL query
try:
`$voice = $question->voices->where('user_id', auth()->id())->first();`

**Comment by Glebsky** [comment]
Good idea to create sendJsonResponse and sendErrorResponse inside base controller so it can be easily use. But laravel has own trick about this. it's called JsonResource.

File: Controller.php (hunk 1 above)
Comment (efalsett) [reply Controller.php:31]: Both would be fine I think, if you split them you have more freedom to change them later on without many if statements

**Pull request closed without merging**